### ISSUE TITLE
using fmp4 causes chromecast gen 2 to freeze

### ISSUE THREAD

**Issue by YouKnowBlom** [jellyfin/jellyfin-chromecast#76]
**Describe the bug**
When hardcoding a transcoding profile to fmp4 over HLS the chromecast locks up completely
```
{
                 Container: "mp4",
                 Type: "Video",
                 AudioCodec: "aac",
                 VideoCodec: "h264",
                 Context: "Streaming",
                 Protocol: "hls",
                 MaxAudioChannels: physicalAudioChannels.toString(),
                 MinSegments: "1",
                 BreakOnNonKeyFrames: false
}
```

**To Reproduce**

1. Hardcode above profile
2. Try to play something that requires transcoding

**Expected behavior**
No freezing and playback works as it should

**System (please complete the following information):**
 - OS: Docker
 - Browser: Chrome
 - Jellyfin Version: 10.7-unstable
 - Cast client: Gen 2

**Comment by nyanmisaka** [comment]
Does chromecast have a native HLS player that support fMP4?
https://developers.google.com/cast/docs/media

hls.js can play fMP4 with h264 AFAIK.

**Comment by YouKnowBlom** [comment]
Yeah it should. They have an example here showing it working here (scroll down just a bit) https://developers.google.com/cast/docs/web_receiver/streaming_protocols#content_protection

**Comment by nyanmisaka** [comment]
Where did it stuck at?
master.m3u8 or main.m3u8? ffmpeg launched or not?

**Comment by YouKnowBlom** [comment]
`master.m3u8`, `main.m3u8` and the first few segments are requested successfully from what I can tell (including `-1.mp4`).

Could it be related to CORS?

**Comment by nyanmisaka** [comment]
> `master.m3u8`, `main.m3u8` and the first few segments are requested successfully from what I can tell (including `-1.mp4`).
> 
> Could it be related to CORS?

I dont think so as TS segments can be requested in the same way.
Might be related to codecs/profile?

**Comment by YouKnowBlom** [comment]
> I dont think so as TS segments can be requested in the same way.
> Might be related to codecs/profile? `H.264 High Profile up to level 4.1 (720p/60fps or 1080p/30fps)`

Yeah that's a good point.

Usually it refuses to play if the level is unsupported, either way there is a limit defined for H.264 in the codecprofiles part of the deviceprofile.

**Comment by nyanmisaka** [comment]
Here are some fMP4 examples from Apple with many different profiles. You can give it a try.
https://developer.apple.com/streaming/examples/

**Comment by YouKnowBlom** [comment]
I'll try a few of those when I get the time. Thanks :p

**Comment by hawken93** [comment]
idittansikte commented 5 hours ago
"I also tried the same thing to enable fmp4 as #76 and played hevc/aac.surr but my CC Ultra froze 60 seconds in (after about 10 segements). But played nicely with sound before that. I had to also set `loadRequestData.media.hlsVideoSegmentFormat = cast.framework.messages.HlsVideoSegmentFormat.FMP4`; in order for it to work at all though."

Just leaving this here :)

**Comment by miquels** [comment]
I ran into the this problem myself today while hacking on one of my projects. Spent hours googling, then created a custom web receiver so that I could debug on the chromecast side. After trying lots of things I think I have managed to solve it on the client (web sender) side. During my googling this github issue also turned up, so I thought I should share. Hope this helps :)

```
    let mediaInfo = new chrome.cast.media.MediaInfo(mediaData.src, 'video/mp4');
    mediaInfo.streamType = chrome.cast.media.StreamType.BUFFERED;

    if (mediaData.src.endsWith(".m3u8")) {
      // These must be set to FMP4, or the chromecast will hang.
      mediaInfo.hlsSegmentFormat = chrome.cast.media.HlsSegmentFormat.FMP4;
      mediaInfo.hlsVideoSegmentFormat = chrome.cast.media.HlsVideoSegmentFormat.FMP4;
      // StreamType must be LIVE or OTHER, not BUFFERED.
      mediaInfo.streamType = chrome.cast.media.StreamType.OTHER;
      // And ofcourse the MIME type.
      mediaInfo.contentType = 'application/x-mpegURL';
    }
```

**Comment by ferferga** [comment]
Care to make a PR? :) @miquels

**Comment by miquels** [comment]
Ha, no thanks. This was just a drive-by comment :).

**Comment by 3flex** [comment]
Anyone willing to retest this using the unstable version? I wonder if this is resolved with the shift to the Shaka Player on Cast devices (and improvements in newer Jellyfin server versions).

**Comment by 3flex** [comment]
Also - [Shaka unconditionally transmuxes TS](https://github.com/shaka-project/shaka-player/pull/6262) so hardcoding fMP4 for Cast might actually make sense (as long as playback is fixed).

I have a Chromecast v2 so may look into that.

**Comment by 3flex** [comment]
Should be fixed by #662. Can test with unstable version.

This seems to be working well on my Chromecast gen 2.
